default/AI/ResourcesAI.py

```diff
@@ -29,98 +38,259 @@
+
+        resource_timer.start("Targets")
+        planet_infos = [PlanetFocusInfo(universe.getPlanet(pid)) for pid in  planet_ids]
+        self.all_planet_info = {}
+        self.all_planet_info.update(zip(planet_ids, planet_infos))
+
+        self.raw_planet_info = dict(self.all_planet_info)
+        self.baked_planet_info = dict()
+
+        for pid, pinfo in self.raw_planet_info.items():
+            focusable = pinfo.planet.currentMeterValue(fo.meterType.targetPopulation) > 0
+            if not focusable:
+                self.baked_planet_info[pid] = self.raw_planet_info.pop(pid)
+
+
+    def bake_future_focus(self, pid, focus, update=True):
+        """Set the focus and moves it from the raw list to the baked list of planets
+        Return success or failure"""
+        pinfo = self.raw_planet_info.get(pid)
+        success = (pinfo is not None and
+                   (pinfo.current_focus == focus
+                    or (focus in pinfo.planet.availableFoci
+                        and fo.issueChangeFocusOrder(pid, focus))))
+        if success:
+            pinfo.future_focus = focus
+            self.baked_planet_info[pid] = self.raw_planet_info.pop(pid)
+
+            if update and pinfo.current_focus != focus:
+                # TODO determine if this should update all planets (for reporting) or just the remaining unplaced planets
+                universe = fo.getUniverse()
+                universe.updateMeterEstimates(self.raw_planet_info.keys())
+                itarget = pinfo.planet.currentMeterValue(fo.meterType.targetIndustry)
+                rtarget = pinfo.planet.currentMeterValue(fo.meterType.targetResearch)
+                pinfo.possible_output[focus] = (itarget, rtarget)
+        return success
+
+
+    def calculate_planet_infos(self, pids):
+        """ Calculates for each possible focus the target output of each planet and stores it in planet info
+        It excludes baked planets from
+        Note: The results will not be strictly correct if any planets have global effects
+        """
+        #TODO this function depends on the specific rule that off-focus meter value are always the minimum value
+        universe = fo.getUniverse()
+        unbaked_pids = [pid for pid in pids if not pid in self.baked_planet_info]
+        planet_infos = [(pid, self.all_planet_info[pid], self.all_planet_info[pid].planet) for pid in unbaked_pids]
+        for pid, pinfo, planet in planet_infos:
+            if IFocus in planet.availableFoci and planet.focus != IFocus:
+                fo.issueChangeFocusOrder(pid, IFocus)  # may not be able to take, but try
+
+        universe.updateMeterEstimates(unbaked_pids)
+        for pid, pinfo, planet in planet_infos:
+            itarget = planet.currentMeterValue(fo.meterType.targetIndustry)
+            rtarget = planet.currentMeterValue(fo.meterType.targetResearch)
+            if planet.focus == IFocus:
+                pinfo.possible_output[IFocus] = (itarget, rtarget)
```

Comment:
`pinfo.possible_output[IFocus]` and `pinfo.possible_output[GFocus]` data collections look pretty familiar for me. But in one case you use list and in other you use tuple.  Please use make them same type for consistency. Tuple is preferred.